src/external/ACSAdapter/AMSFileScanner.ts

```diff
@@ -0,0 +1,162 @@
+        this.scanPollingInterval = scanPollingInterval;
+        this.shouldQueueScan = false;
+        this.test = false;
+        this.queueScan();
+    }
+
+    public async queueScan(): Promise<void> {
+        this.shouldQueueScan = true;
+        try {
+            await this.scanFiles();
+        } catch (e) {
+            console.error(e);
+        } finally {
+            await WebUtils.sleep(this.scanPollingInterval);
+            this.shouldQueueScan && !this.test && await this.queueScan();
+        }
+    }
+
+    public retrieveFileScanResult(id: string): FileScanResult | undefined {
+        return this.scanResults?.get(id);
+    }
+
+    public addOrUpdateFile(id: string, fileMetadata: FileMetadata, scan: FileScanResponse): void {
+        const scanResult = this.retrieveFileScanResult(id);
+        this.scanResults?.set(id, {
+            ...(scanResult || {}),
+            fileMetadata,
+            scan
+        });
+    }
+
+    public addNext(id: string, next: Function): void {
+        const scanResult = this.scanResults?.get(id);
+        if (scanResult) {
+            this.scanResults?.set(id, {...scanResult, next});
+        }
+    }
+
+    public addActivity(id: string, activity: any): void {  // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+        const scanResult = this.scanResults?.get(id);
+        if (scanResult) {
+            this.scanResults?.set(id, {...scanResult, activity});
+        }
+    }
+
+    public async scanFileCallback(scanResult: FileScanResult, id: string): Promise<void> {
+        const {fileMetadata, next, activity, scan} = scanResult;
+
+        if (scan?.status === AMSViewScanStatus.INPROGRESS) {
+            try {
+                const response: any = await this.amsClient.getViewStatus(fileMetadata); // eslint-disable-line @typescript-eslint/no-explicit-any
+                const {view_location, scan} = response;
+
+                this.addOrUpdateFile(id, scanResult.fileMetadata, scan);
+
+                if (scan.status === AMSViewScanStatus.PASSED && next && activity) {
+                   await this.renderAttachmentActivity(scanResult, view_location);
+                }
+
+                if (scan.status === AMSViewScanStatus.MALWARE && next && activity) {
+                    await this.renderMalwareActivity(scanResult);
+                }
+            } catch (e) {
+                console.error(e);
+            }
+
+            await WebUtils.sleep(1000);
+        }
+    }
+
+    public scanFiles(): Promise<void> {
+        this.scanResults?.forEach(async (scanResult, id) => {await this.scanFileCallback(scanResult, id)});
+
+        // eslint-disable-next-line no-async-promise-executor
+        return new Promise(async (resolve) => {
+            this.scanResults?.forEach(async (scanResult, id) => {await this.scanFileCallback(scanResult, id)});
+            resolve();
+        });

```

Comment:
Slightly confused by the code execution here -- will we want to be running `scanResults.forEach(scanFileCallback())` multiple times / twice?

Reply:
Good catch. It's been removed.